toolchain/semantics/semantics_ir_factory.cpp

```diff
@@ -7,155 +7,55 @@
 #include <stack>
 
 #include "common/check.h"
-#include "llvm/ADT/StringMap.h"
-#include "llvm/Support/FormatVariadic.h"
 #include "toolchain/lexer/token_kind.h"
 #include "toolchain/lexer/tokenized_buffer.h"
 #include "toolchain/parser/parse_node_kind.h"
-//#include "toolchain/semantics/meta_node_block.h"
+#include "toolchain/semantics/node_ref.h"
 #include "toolchain/semantics/nodes/binary_operator.h"
-#include "toolchain/semantics/parse_subtree_consumer.h"
+#include "toolchain/semantics/nodes/integer_literal.h"
 
 namespace Carbon {
 
-// The ParseTree is walked in reverse post order, meaning a lot of nodes are
-// added in reverse. This fixes that ordering to be the easier to understand
-// code ordering.
-template <typename T>
-static void FixReverseOrdering(T& container) {
-  std::reverse(container.begin(), container.end());
-}
-
 auto SemanticsIRFactory::Build(const TokenizedBuffer& tokens,
                                const ParseTree& parse_tree) -> SemanticsIR {
   SemanticsIRFactory builder(tokens, parse_tree);
   builder.Build();
   return builder.semantics_;
 }
 
-void SemanticsIRFactory::Build() {
-  auto subtree = ParseSubtreeConsumer::ForTree(parse_tree());
-  semantics_.root_block_ =
-      TransformBlockSubtree(subtree, ParseNodeKind::FileEnd());
-}
-
-void SemanticsIRFactory::RequireNodeEmpty(ParseTree::Node node) {
-  auto subtree_size = parse_tree().node_subtree_size(node);
+// An entry in the stack for traversing the ParseTree.
+// TODO: This is badly structured, and a redesign may be able to get rid of
+// SemanticsIR. Need to keep thinking about this, but for now this setup allows
+// for test consistency.
+struct TraversalStackEntry {
+  explicit TraversalStackEntry(ParseTree::Node parse_node)
+      : parse_node(parse_node) {}
+
+  TraversalStackEntry(ParseTree::Node parse_node,
+                      llvm::SmallVector<Semantics::NodeRef, 0> sem_ir)
+      : parse_node(parse_node), sem_ir(std::move(sem_ir)) {}
+
+  TraversalStackEntry(ParseTree::Node parse_node,
+                      llvm::SmallVector<Semantics::NodeRef, 0> sem_ir,
+                      Semantics::NodeId result_id)
+      : parse_node(parse_node),
+        sem_ir(std::move(sem_ir)),
+        result_id(result_id) {}
+
+  ParseTree::Node parse_node;
+  llvm::SmallVector<Semantics::NodeRef, 0> sem_ir;
```

Comment:
Having each stack entry have its own separately-allocated vector seems inefficient to me -- I think it will lead to a bunch of avoidable heap allocations. I wonder if it would be better to have a single central vector of nodes, and have each stack entry hold a pair of indexes to the first and (one-past) last node that it owns.

Reply:
Agreed, this was the above todo, I may be thinking a different direction though. That is, maybe I can just print most of sem_ir in-order (for the most part).